Add List table view cell component

dvdchr opened the pull request:
Refs #16741

## Summary
This introduces the table view cell for the List component. Some notes regarding this change:

- The cells are still detached from the application flow, and individual adjustments to Comments list and Notifications list will be made in separate PRs.
- Indicator views (the little colored dot on the left) has a default color of `.clear`. This is because both Notifications list assigns blue color for unread state, while Comments list assigns yellow color for unmoderated state. Therefore, it's best for the users of the List component to manage their own indicator colors (will be added in later PRs).
- The number of lines for `titleLabel` is increased to 3 when snippet text is empty or nil.

Here's how the cells look in Notifications screen (note that `ListTableHeaderView` is also used here):

Light 🌝  | Dark 🌚 
-- | --
![16741_light_noSnippet](https://user-images.githubusercontent.com/1299411/125279619-5b0d2e00-e33e-11eb-9da7-699283713e3d.png) | ![16741_dark_noSnippet](https://user-images.githubusercontent.com/1299411/125279607-58123d80-e33e-11eb-85f1-9929af9b443b.png)

## To Test
Since the table view cell is still detached from application flow, you'd need to manually register the table view cells in either Comments list (`CommentsViewController`) or Notifications list (`NotificationsViewController`).

## Regression Notes
1. Potential unintended areas of impact
None. The view is detached from application flow.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the component in `NotificationsViewController`.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Add snippet table view cell

dvdchr committed:
Add cells for List component

dvdchr committed:
Refactor SnippetTableViewCell to ListTableViewCell

dvdchr committed:
Minor tweak cell to match design

dvdchr committed:
Add some comments for ListTableViewCell

dvdchr committed:
Show more titleLabel lines when there's no snippet

dvdchr committed:
Add some more comments, clean up

dvdchr committed:
Adjust top spacing to match design

dvdchr commented:
Looping in @mattmiklic to optionally review the designs.

In addition, I've noticed based on the designs that there's a 12pt top padding for the avatar image, but it's only 10pt for the title label. In the attached screenshots, the top padding for the avatar image is still 10pt – but this has been addressed in [dfeb24b](https://github.com/wordpress-mobile/WordPress-iOS/pull/16843/commits/dfeb24bdd6bbcad18854640996a98b9bfc8e217b).

aerych commented on WordPress/Classes/ViewRelated/Views/List/ListTableViewCell.swift:
```diff
@@ -0,0 +1,122 @@
+        if let someURL = url, let gravatar = Gravatar(someURL) {
+            avatarView.downloadGravatar(gravatar, placeholder: placeholderImage, animate: true) { error in
+                if let error = error {
+                    print(error)
```
Rather than `print` it would be nice to use `DDLogError` along with a short description to give context to where the error occurred. This way the error is recorded to the activity log and we have the option of reviewing it should our user contact support about an issue they're having.

dvdchr replied:
Oof! This was added for personal debugging, and I forgot to take it out. Thank you for noticing! I'll remove this instead.

aerych commented on WordPress/Classes/ViewRelated/Views/List/ListTableViewCell.swift:
```diff
@@ -0,0 +1,122 @@
+    // MARK: Properties
+
+    /// The color of the indicator circle.
+    @objc var indicatorColor: UIColor = .clear
```
Should this have a `didSet` to update `indicatorView.backgroundColor` similar to `showsIndicator`?  Without it, the order that `indicatorColor` and `showsIndicator` are called matters.  Call them in the wrong order and the color doesn't update until the next time the cell is refreshed.

dvdchr replied:
Good catch, sir. I'll adjust it soon!

aerych commented on WordPress/Classes/ViewRelated/Views/List/ListTableViewCell.swift:
```diff
@@ -0,0 +1,122 @@
+
+    /// Show more lines in titleLabel when there's no snippet.
+    func updateTitleTextLines() {
+        let hasSnippet = !(snippetText ?? "").isEmpty
```
Just a thought, but I could see this being it's own method.

aerych reviewed (changes requested):
Heya @dvdchr 👋 
Looking good.  I had a couple of minor questions/suggestions.  Lemme know what you think!

Is the next PR to wire up the new UI behind a feature flag?  It'll be nice to see everything working all together without needing to make manual edits. :)

mattmiklic commented:
Since I can't see the changes in a test build I'm just going by the screenshots. But everything looks like it's in order here to me. We can address any design issues once these are live in the app. 👍

dvdchr commented:
> Is the next PR to wire up the new UI behind a feature flag?

Yup! Next step is to add a feature flag, and then integrate it to Comments and Notifications.

> We can address any design issues once these are live in the app.

Noted, thank you!

dvdchr committed:
Adjustments for ListTableViewCell

- add `didSet` to `indicatorColor` property that updates the color of indicatorView.
- extracted `hasSnippet` into a computed property.
- removed debug print method.

dvdchr committed:
Merge branch 'develop' into issue/16741-list-table-cells

aerych reviewed (approved):
Looks good! :shipit: sir!
